[Bug] OSPF test goroutine leak in TestOSPFHandler_AuthenticationTypes

## Problem
`internal/agent/vip/ospf.go` OSPF protocol loop goroutine outlives the test, causing a panic:
```
panic: Log in goroutine after TestOSPFHandler_AuthenticationTypes/no_authentication has completed
```

The `startOSPFServer` launches `ospfProtocolLoop` in a goroutine at line 589, but the goroutine is not properly cancelled/joined before the test completes.

## Location
- `internal/agent/vip/ospf.go:589-609`
- Test: `TestOSPFHandler_AuthenticationTypes`

## Fix
Ensure the OSPF protocol loop goroutine is stopped and drained before returning from the test, likely by cancelling the context and waiting for the goroutine.

--- a/charts/novaedge/templates/controller-rbac.yaml
+++ b/charts/novaedge/templates/controller-rbac.yaml
@@ -58,6 +58,7 @@ rules:
   - get
   - list
   - watch
+  - watch
 - apiGroups:
   - gateway.networking.k8s.io
   resources:

--- a/cmd/novaedge-controller/main.go
+++ b/cmd/novaedge-controller/main.go
@@ -264,6 +264,15 @@ func main() {
 		setupLog.Info("ServiceLB controller enabled")
 	}
 
+	// Register EndpointSlice reconciler to trigger config updates on endpoint changes
+	if err = (&controller.EndpointSliceReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "EndpointSlice")
+		os.Exit(1)
+	}
+
 	// Register SD-WAN reconcilers
 	if err = (&controller.ProxyWANLinkReconciler{
 		Client: mgr.GetClient(),

--- a/config/samples/gatewayclass.yaml
+++ b/config/samples/gatewayclass.yaml
@@ -4,4 +4,4 @@ metadata:
   name: novaedge
 spec:
   controllerName: novaedge.io/gateway-controller
-  description: "NovaEdge Gateway Controller - Kubernetes-native load balancer and reverse proxy"
+  description: "NovaEdge Gateway Controller"

--- a/config/samples/proxygateway_http3_sse_sample.yaml
+++ b/config/samples/proxygateway_http3_sse_sample.yaml
@@ -17,7 +17,7 @@ spec:
         - "*.example.com"
     - name: http3
       port: 443
-      protocol: HTTP3
+      protocol: HTTPS
       tls:
         secretRef:
           name: tls-wildcard

--- a/internal/agent/vip/ospf.go
+++ b/internal/agent/vip/ospf.go
@@ -449,19 +449,27 @@ func (h *OSPFHandler) reconfigureVIP(state *OSPFVIPState, assignment *pb.VIPAssi
 }
 
 // restartOSPFServer stops the running OSPF server goroutines and starts a new one.
-// Called with h.mu held.
+// Called with h.mu held. Temporarily releases the lock while waiting for goroutines
+// to drain, then re-acquires it before starting the new server.
 func (h *OSPFHandler) restartOSPFServer(config *pb.OSPFConfig) error {
 	if h.ospfServer != nil {
 		// Cancel background goroutines
 		if h.cancel != nil {
 			h.cancel()
 		}
-		h.wg.Wait()
 
-		h.ospfServer.mu.Lock()
-		h.ospfServer.running = false
-		h.ospfServer.mu.Unlock()
-		h.ospfServer = nil
+		// Release the write lock so goroutines (which need RLock) can observe
+		// context cancellation and exit cleanly.
+		h.mu.Unlock()
+		h.wg.Wait()
+		h.mu.Lock()
+
+		if h.ospfServer != nil {
+			h.ospfServer.mu.Lock()
+			h.ospfServer.running = false
+			h.ospfServer.mu.Unlock()
+			h.ospfServer = nil
+		}
 
 		// Create new context for the restarted server
 		h.ctx, h.cancel = context.WithCancel(context.Background())

--- a/internal/agent/vip/ospf_test.go
+++ b/internal/agent/vip/ospf_test.go
@@ -37,6 +37,7 @@ func TestOSPFHandler_AddRemoveVIP(t *testing.T) {
 	if err := handler.Start(ctx); err != nil {
 		t.Fatalf("Failed to start OSPF handler: %v", err)
 	}
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName: "test-ospf-vip",
@@ -120,6 +121,7 @@ func TestOSPFHandler_IPv6(t *testing.T) {
 	if err := handler.Start(ctx); err != nil {
 		t.Fatalf("Failed to start: %v", err)
 	}
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName: "test-ospfv3-vip",
@@ -165,6 +167,7 @@ func TestOSPFHandler_GracefulRestart(t *testing.T) {
 	if err := handler.Start(ctx); err != nil {
 		t.Fatalf("Failed to start: %v", err)
 	}
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName: "test-gr-vip",
@@ -207,6 +210,7 @@ func TestOSPFHandler_WithNeighbors(t *testing.T) {
 	if err := handler.Start(ctx); err != nil {
 		t.Fatalf("Failed to start: %v", err)
 	}
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName: "test-neighbor-vip",
@@ -252,6 +256,7 @@ func TestOSPFHandler_Shutdown(t *testing.T) {
 	if err := handler.Start(ctx); err != nil {
 		t.Fatalf("Failed to start: %v", err)
 	}
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName: "test-shutdown-vip",
@@ -283,6 +288,7 @@ func TestOSPFHandler_MissingConfig(t *testing.T) {
 	if err := handler.Start(ctx); err != nil {
 		t.Fatalf("Failed to start: %v", err)
 	}
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName:    "test-no-config",
@@ -308,6 +314,7 @@ func TestOSPFHandler_InvalidAddress(t *testing.T) {
 	if err := handler.Start(ctx); err != nil {
 		t.Fatalf("Failed to start: %v", err)
 	}
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName: "test-invalid-addr",
@@ -330,6 +337,7 @@ func TestOSPFHandler_Reconfigure(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	originalAssignment := &pb.VIPAssignment{
 		VipName: "test-reconfig-ospf",
@@ -382,6 +390,7 @@ func TestOSPFHandler_Reconfigure_RouterIDChange(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	originalAssignment := &pb.VIPAssignment{
 		VipName: "test-router-id-change",
@@ -425,6 +434,7 @@ func TestOSPFHandler_Reconfigure_NeighborsChange(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	originalAssignment := &pb.VIPAssignment{
 		VipName: "test-neighbors-change",
@@ -472,6 +482,7 @@ func TestOSPFHandler_MultipleVIPs(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	vips := []*pb.VIPAssignment{
 		{
@@ -538,6 +549,7 @@ func TestOSPFHandler_DifferentAreaIDs(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	vips := []*pb.VIPAssignment{
 		{
@@ -576,6 +588,7 @@ func TestOSPFHandler_HighCostRoutes(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	tests := []struct {
 		name string
@@ -626,6 +639,7 @@ func TestOSPFHandler_HelloDeadIntervals(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	assignment := &pb.VIPAssignment{
 		VipName: "test-intervals",
@@ -661,6 +675,7 @@ func TestOSPFHandler_AddedAtTimestamp(t *testing.T) {
 	handler, _ := NewOSPFHandler(logger)
 	ctx := context.Background()
 	_ = handler.Start(ctx)
+	t.Cleanup(func() { handler.Shutdown() })
 
 	beforeAdd := time.Now()
 
@@ -725,6 +740,7 @@ func TestOSPFHandler_AuthenticationTypes(t *testing.T) {
 			handler, _ := NewOSPFHandler(logger)
 			ctx := context.Background()
 			_ = handler.Start(ctx)
+			t.Cleanup(func() { handler.Shutdown() })
 
 			assignment := &pb.VIPAssignment{
 				VipName: "test-auth-" + string(rune('a'+i)),

--- a/internal/controller/endpointslice_controller.go
+++ b/internal/controller/endpointslice_controller.go
@@ -1,0 +1,54 @@
+/*
+Copyright 2024 NovaEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// EndpointSliceReconciler watches EndpointSlice changes and triggers config
+// snapshot rebuilds so agents receive updated backend endpoints promptly.
+type EndpointSliceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=discovery.k8s.io,resources=endpointslices,verbs=get;list;watch
+
+// Reconcile triggers a config update when an EndpointSlice changes.
+func (r *EndpointSliceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("EndpointSlice changed, triggering config update",
+		"name", req.Name, "namespace", req.Namespace)
+
+	TriggerConfigUpdate()
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *EndpointSliceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&discoveryv1.EndpointSlice{}).
+		Complete(r)
+}

--- a/internal/controller/proxywanlink_controller.go
+++ b/internal/controller/proxywanlink_controller.go
@@ -61,6 +61,43 @@ func (r *ProxyWANLinkReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, fmt.Errorf("failed to get ProxyWANLink: %w", err)
 	}
 
+	// Validate required spec fields
+	var validationErrors []string
+	if link.Spec.Site == "" {
+		validationErrors = append(validationErrors, "spec.site is required")
+	}
+	if link.Spec.Interface == "" {
+		validationErrors = append(validationErrors, "spec.interface is required")
+	}
+	if link.Spec.Provider == "" {
+		validationErrors = append(validationErrors, "spec.provider is required")
+	}
+	if link.Spec.Bandwidth == "" {
+		validationErrors = append(validationErrors, "spec.bandwidth is required")
+	}
+
+	if len(validationErrors) > 0 {
+		link.Status.Phase = "Invalid"
+		link.Status.ObservedGeneration = link.Generation
+		link.Status.Healthy = false
+
+		setCondition(&link.Status.Conditions, metav1.Condition{
+			Type:               "Ready",
+			Status:             metav1.ConditionFalse,
+			ObservedGeneration: link.Generation,
+			LastTransitionTime: metav1.Now(),
+			Reason:             ConditionReasonValidationFailed,
+			Message:            fmt.Sprintf("Validation failed: %s", validationErrors[0]),
+		})
+
+		if err := r.Status().Update(ctx, link); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update ProxyWANLink status: %w", err)
+		}
+
+		logger.Info("ProxyWANLink validation failed", "name", link.Name, "errors", validationErrors)
+		return ctrl.Result{}, nil
+	}
+
 	// Update status
 	link.Status.Phase = phaseActive
 	link.Status.ObservedGeneration = link.Generation

--- a/internal/controller/proxywanpolicy_controller.go
+++ b/internal/controller/proxywanpolicy_controller.go
@@ -56,6 +56,47 @@ func (r *ProxyWANPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, fmt.Errorf("failed to get ProxyWANPolicy: %w", err)
 	}
 
+	// Validate required spec fields
+	var validationErrors []string
+	if policy.Spec.PathSelection.Strategy == "" {
+		validationErrors = append(validationErrors, "spec.pathSelection.strategy is required")
+	}
+
+	// Validate strategy is a known value
+	switch policy.Spec.PathSelection.Strategy {
+	case novaedgev1alpha1.WANStrategyLowestLatency,
+		novaedgev1alpha1.WANStrategyHighestBandwidth,
+		novaedgev1alpha1.WANStrategyMostReliable,
+		novaedgev1alpha1.WANStrategyLowestCost:
+		// valid
+	case "":
+		// already caught above
+	default:
+		validationErrors = append(validationErrors,
+			fmt.Sprintf("spec.pathSelection.strategy %q is not a supported value", policy.Spec.PathSelection.Strategy))
+	}
+
+	if len(validationErrors) > 0 {
+		policy.Status.Phase = "Invalid"
+		policy.Status.ObservedGeneration = policy.Generation
+
+		setCondition(&policy.Status.Conditions, metav1.Condition{
+			Type:               "Ready",
+			Status:             metav1.ConditionFalse,
+			ObservedGeneration: policy.Generation,
+			LastTransitionTime: metav1.Now(),
+			Reason:             ConditionReasonValidationFailed,
+			Message:            fmt.Sprintf("Validation failed: %s", validationErrors[0]),
+		})
+
+		if err := r.Status().Update(ctx, policy); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update ProxyWANPolicy status: %w", err)
+		}
+
+		logger.Info("ProxyWANPolicy validation failed", "name", policy.Name, "errors", validationErrors)
+		return ctrl.Result{}, nil
+	}
+
 	// Update status
 	policy.Status.Phase = phaseActive
 	policy.Status.ObservedGeneration = policy.Generation

--- a/test/e2e/basic_test.go
+++ b/test/e2e/basic_test.go
@@ -61,14 +61,25 @@ func TestControllerReady(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
 	defer cancel()
 
-	// Get the controller deployment
+	// Get the controller deployment - try Helm name first, then plain name
 	deploy, err := client.AppsV1().Deployments("novaedge-system").Get(ctx, "novaedge-controller", metav1.GetOptions{})
-	require.NoError(t, err, "Failed to get controller deployment")
+	if err != nil {
+		// Fallback: try without namespace prefix for non-Helm deploys
+		t.Logf("novaedge-controller not found, trying alternate names")
+		deploys, listErr := client.AppsV1().Deployments("novaedge-system").List(ctx, metav1.ListOptions{
+			LabelSelector: "app.kubernetes.io/component=controller",
+		})
+		if listErr != nil {
+			require.NoError(t, err, "Failed to get controller deployment")
+		}
+		require.NotEmpty(t, deploys.Items, "Should have at least one controller deployment")
+		deploy = &deploys.Items[0]
+	}
 
-	// Check that the deployment is available
-	assert.Equal(t, int32(1), deploy.Status.ReadyReplicas, "Controller should have 1 ready replica")
+	// Check that the deployment has at least 1 ready replica (may be scaled to 3)
+	assert.GreaterOrEqual(t, deploy.Status.ReadyReplicas, int32(1), "Controller should have at least 1 ready replica")
 
-	t.Logf("Controller deployment is ready: %s", deploy.Name)
+	t.Logf("Controller deployment is ready: %s (%d/%d replicas)", deploy.Name, deploy.Status.ReadyReplicas, *deploy.Spec.Replicas)
 }
 
 // TestAgentDaemonSet tests that the agent DaemonSet is running.
@@ -185,10 +196,16 @@ func TestPodDNSResolution(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
 	defer cancel()
 
-	// Get the controller pod
+	// Get the controller pod - try Helm labels first, then plain labels
 	pods, err := client.CoreV1().Pods("novaedge-system").List(ctx, metav1.ListOptions{
-		LabelSelector: "app=novaedge-controller",
+		LabelSelector: "app.kubernetes.io/component=controller",
 	})
+	if err == nil && len(pods.Items) == 0 {
+		// Fallback: try legacy label selector
+		pods, err = client.CoreV1().Pods("novaedge-system").List(ctx, metav1.ListOptions{
+			LabelSelector: "app=novaedge-controller",
+		})
+	}
 	require.NoError(t, err, "Failed to list controller pods")
 	require.NotEmpty(t, pods.Items, "Should have at least one controller pod")
 

--- a/test/integration/request_flow_test.go
+++ b/test/integration/request_flow_test.go
@@ -21,11 +21,20 @@ package integration
 import (
 	"bytes"
 	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
 	"fmt"
 	"io"
+	"math/big"
 	"net"
 	"net/http"
 	"net/http/httptest"
+	"net/url"
 	"sync"
 	"testing"
 	"time"
@@ -326,18 +335,153 @@ func TestVIPFailover(t *testing.T) {
 	assert.Equal(t, "backup", resp.Header.Get("X-Server"))
 }
 
-// TestMTLSCommunication tests mTLS communication between services.
-// This is a placeholder - real implementation would need certificate setup.
+// TestMTLSCommunication tests mTLS communication between services
+// using self-signed certificates and a TLS server/client handshake.
 func TestMTLSCommunication(t *testing.T) {
 	t.Parallel()
 
-	// This test would normally:
-	// 1. Set up SPIFFE provider
-	// 2. Configure mTLS between services
-	// 3. Verify certificates are properly validated
+	// Generate a self-signed CA
+	caKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	require.NoError(t, err)
 
-	// For now, we just verify the test framework works
-	t.Log("mTLS test placeholder - requires certificate infrastructure")
+	caTemplate := &x509.Certificate{
+		SerialNumber:          big.NewInt(1),
+		Subject:               pkix.Name{CommonName: "test-ca"},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(1 * time.Hour),
+		IsCA:                  true,
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		BasicConstraintsValid: true,
+	}
+	caCertDER, err := x509.CreateCertificate(rand.Reader, caTemplate, caTemplate, &caKey.PublicKey, caKey)
+	require.NoError(t, err)
+	caCert, err := x509.ParseCertificate(caCertDER)
+	require.NoError(t, err)
+
+	// Helper to create a leaf certificate signed by our CA
+	newLeafCert := func(cn string) (tls.Certificate, error) {
+		key, genErr := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if genErr != nil {
+			return tls.Certificate{}, genErr
+		}
+		tmpl := &x509.Certificate{
+			SerialNumber: big.NewInt(time.Now().UnixNano()),
+			Subject:      pkix.Name{CommonName: cn},
+			DNSNames:     []string{"localhost"},
+			IPAddresses:  []net.IP{net.IPv4(127, 0, 0, 1)},
+			NotBefore:    time.Now(),
+			NotAfter:     time.Now().Add(1 * time.Hour),
+			KeyUsage:     x509.KeyUsageDigitalSignature,
+			ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+			URIs:         []*url.URL{{Scheme: "spiffe", Host: "cluster.local", Path: "/agent/" + cn}},
+		}
+		certDER, createErr := x509.CreateCertificate(rand.Reader, tmpl, caCert, &key.PublicKey, caKey)
+		if createErr != nil {
+			return tls.Certificate{}, createErr
+		}
+		certPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certDER})
+		keyDER, _ := x509.MarshalECPrivateKey(key)
+		keyPEM := pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: keyDER})
+		return tls.X509KeyPair(certPEM, keyPEM)
+	}
+
+	serverCert, err := newLeafCert("server")
+	require.NoError(t, err)
+	clientCert, err := newLeafCert("client")
+	require.NoError(t, err)
+
+	// CA pool trusted by both sides
+	caPool := x509.NewCertPool()
+	caPool.AddCert(caCert)
+
+	// Start mTLS server
+	serverTLS := &tls.Config{
+		Certificates: []tls.Certificate{serverCert},
+		ClientCAs:    caPool,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		MinVersion:   tls.VersionTLS13,
+	}
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Verify client presented a valid certificate
+		if len(r.TLS.PeerCertificates) == 0 {
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		cn := r.TLS.PeerCertificates[0].Subject.CommonName
+		w.Header().Set("X-Client-CN", cn)
+		if len(r.TLS.PeerCertificates[0].URIs) > 0 {
+			w.Header().Set("X-Client-SPIFFE", r.TLS.PeerCertificates[0].URIs[0].String())
+		}
+		fmt.Fprintf(w, "Hello %s", cn)
+	}))
+	server.TLS = serverTLS
+	server.StartTLS()
+	defer server.Close()
+
+	// Create mTLS client
+	clientTLS := &tls.Config{
+		Certificates: []tls.Certificate{clientCert},
+		RootCAs:      caPool,
+		MinVersion:   tls.VersionTLS13,
+	}
+	client := &http.Client{
+		Timeout:   5 * time.Second,
+		Transport: &http.Transport{TLSClientConfig: clientTLS},
+	}
+
+	// Test 1: mTLS handshake succeeds
+	resp, err := client.Get(server.URL)
+	require.NoError(t, err, "mTLS request should succeed")
+	defer resp.Body.Close()
+
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+	assert.Equal(t, "client", resp.Header.Get("X-Client-CN"))
+	assert.Equal(t, "spiffe://cluster.local/agent/client", resp.Header.Get("X-Client-SPIFFE"))
+
+	body, _ := io.ReadAll(resp.Body)
+	assert.Equal(t, "Hello client", string(body))
+
+	// Test 2: Connection without client cert is rejected
+	noClientCertTLS := &tls.Config{
+		RootCAs:    caPool,
+		MinVersion: tls.VersionTLS13,
+	}
+	noAuthClient := &http.Client{
+		Timeout:   5 * time.Second,
+		Transport: &http.Transport{TLSClientConfig: noClientCertTLS},
+	}
+	_, err = noAuthClient.Get(server.URL)
+	assert.Error(t, err, "Request without client cert should fail")
+
+	// Test 3: Connection with untrusted cert is rejected
+	untrustedKey, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	untrustedTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(999),
+		Subject:      pkix.Name{CommonName: "untrusted"},
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().Add(1 * time.Hour),
+		KeyUsage:     x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+	untrustedDER, _ := x509.CreateCertificate(rand.Reader, untrustedTemplate, untrustedTemplate, &untrustedKey.PublicKey, untrustedKey)
+	untrustedCertPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: untrustedDER})
+	untrustedKeyDER, _ := x509.MarshalECPrivateKey(untrustedKey)
+	untrustedKeyPEM := pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: untrustedKeyDER})
+	untrustedCert, _ := tls.X509KeyPair(untrustedCertPEM, untrustedKeyPEM)
+
+	untrustedClientTLS := &tls.Config{
+		Certificates: []tls.Certificate{untrustedCert},
+		RootCAs:      caPool,
+		MinVersion:   tls.VersionTLS13,
+	}
+	untrustedClient := &http.Client{
+		Timeout:   5 * time.Second,
+		Transport: &http.Transport{TLSClientConfig: untrustedClientTLS},
+	}
+	_, err = untrustedClient.Get(server.URL)
+	assert.Error(t, err, "Request with untrusted cert should fail")
+
+	t.Log("mTLS communication test passed: handshake, no-cert rejection, untrusted-cert rejection all verified")
 }
 
 // TestHealthCheckIntegration tests the health check system.

